### ISSUE TITLE
[PR #1214] fix(modkit-db, mini-chat): handle unique-constraint violations for vector stores

### ISSUE THREAD

**Issue by orivaris** [constructorfabric/gears-rust#1670]
> 🔗 **Mirrored PR** [cyberfabric/cyberware-rust#1214](https://github.com/cyberfabric/cyberware-rust/pull/1214) | **Author:** MikeFalcon77 | **Opened:** 2026-03-17T23:19:44Z | **Status:** merged on 2026-03-18T08:03:26Z
> *GitHub API does not allow setting PR author or timestamps — attribution preserved here.*

---

- Add `is_unique_violation()` to modkit-db
- Use it in mini-chat's `map_db_err` so the vector-store race on duplicate file uploads returns 409 Conflict instead of 500 Internal Error

## Summary by CodeRabbit

* **Improvements**
  * Better detection of database unique-constraint violations across Postgres, SQLite, and MySQL, producing clearer "conflict" responses when duplicates are detected.
  * More resilient handling of concurrent vector-store creation: insert races are detected and resolved via fallback checks, polling, and warnings to recover existing assignments and avoid spurious failures.

---

**Comment by orivaris** [comment]
**coderabbitai[bot]** commented on 2026-03-17T23:19:59Z:

---




<details>
<summary>📝 Walkthrough</summary>

## Walkthrough

Adds a public utility to detect SeaORM unique-constraint violations, re-exports it, integrates detection into DB→domain error mapping, and adds a race-aware handler for concurrent vector-store inserts that polls or returns existing IDs when conflicts occur.

## Changes

|Cohort / File(s)|Summary|
|---|---|
|**Unique Violation Detection Utility** <br> `libs/modkit-db/src/secure/error.rs`, `libs/modkit-db/src/secure/mod.rs`|Adds `pub fn is_unique_violation(&sea_orm::DbErr) -> bool` and `ScopeError::is_unique_violation(&self) -> bool`; detection uses `sql_err()` with lowercase-message fallbacks for Postgres/SQLite/MySQL. Re-exports `is_unique_violation` from the `secure` module.|
|**Error Mapping Integration** <br> `modules/mini-chat/mini-chat/src/domain/error.rs`|`map_db_err()` now calls `modkit_db::secure::is_unique_violation()` as a fallback and maps detected unique-constraint violations to `DomainError::Conflict` with code `"unique_violation"`.|
|**Race Condition Handler** <br> `modules/mini-chat/mini-chat/src/domain/service/attachment_service.rs`|Replaces direct insert error propagation with `handle_vector_store_insert_race()` which: checks for existing row, returns existing `vector_store_id`, polls if present but NULL, or re-raises the original DB error. Adds logging for concurrent-insert scenarios.|

## Sequence Diagram(s)

```mermaid
sequenceDiagram
    participant Caller
    participant AttachmentService
    participant Database
    participant Poller

    Caller->>AttachmentService: get_or_create_vector_store(...)
    AttachmentService->>Database: INSERT INTO vector_store(...)
    Database-->>AttachmentService: DbErr (unique constraint)
    AttachmentService->>AttachmentService: handle_vector_store_insert_race(original_err)
    AttachmentService->>Database: SELECT vector_store_id FROM vector_store WHERE keys
    alt row exists with non-null vector_store_id
        Database-->>AttachmentService: row + vector_store_id
        AttachmentService-->>Caller: return existing vector_store_id
    else row exists with NULL vector_store_id
        Database-->>AttachmentService: row with NULL
        AttachmentService->>Poller: poll_vector_store_until_populated(...)
        Poller->>Database: repeated SELECTs
        Database-->>Poller: row now has vector_store_id
        Poller-->>AttachmentService: return populated vector_store_id
        AttachmentService-->>Caller: return populated vector_store_id
    else no row found
        Database-->>AttachmentService: no row
        AttachmentService-->>Caller: re-raise original_err
    end
```

## Estimated code review effort

🎯 3 (Moderate) | ⏱️ ~22 minutes

## Possibly related issues

- constructorfabric/cyberfabric-core#1215 — Adds the same unique-constraint detection and handling logic and the mini-chat insert-race handling described in this PR.

## Poem

> 🐰 I sniffed a race on fields so tight,  
> > A duplicate key gave quite a fright,  
> > I peeked and polled until the thread was clear,  
> > Found the winner, returned its ID near —  
> > Hooray, the rows hopped back in sight! 🎉

</details>






<details>
<summary>🚥 Pre-merge checks | ✅ 3</summary>

<details>
<summary>✅ Passed checks (3 passed)</summary>

|     Check name     | Status   | Explanation                                                                                                                                               |
| :----------------: | :------- | :-------------------------------------------------------------------------------------------------------------------------------------------------------- |
|  Description Check | ✅ Passed | Check skipped - CodeRabbit’s high-level summary is enabled.                                                                                               |
|     Title check    | ✅ Passed | The title accurately summarizes the main change: adding unique-constraint violation handling for vector stores across modkit-db and mini-chat components. |
| Docstring Coverage | ✅ Passed | Docstring coverage is 100.00% which is sufficient. The required threshold is 80.00%.                                                                      |

</details>

<sub>✏️ Tip: You can configure your own custom pre-merge checks in the settings.</sub>

</details>





<details>
<summary>✨ Finishing Touches</summary>

<details>
<summary>🧪 Generate unit tests (beta)</summary>

- [ ]    Create PR with unit tests

</details>

</details>





<details>
<summary>📝 Coding Plan</summary>

- [ ]  Generate coding plan for human review comments

</details>





---

Thanks for using [CodeRabbit](https://coderabbit.ai?utm_source=oss&utm_medium=github&utm_campaign=cyberfabric/cyberfabric-core&utm_content=1214)! It's free for OSS, and your support helps us grow. If you like it, consider giving us a shout-out.

<details>
<summary>❤️ Share</summary>

- [X](https://twitter.com/intent/tweet?text=I%20just%20used%20%40coderabbitai%20for%20my%20code%20review%2C%20and%20it%27s%20fantastic%21%20It%27s%20free%20for%20OSS%20and%20offers%20a%20free%20trial%20for%20the%20proprietary%20code.%20Check%20it%20out%3A&url=https%3A//coderabbit.ai)
- [Mastodon](https://mastodon.social/share?text=I%20just%20used%20%40coderabbitai%20for%20my%20code%20review%2C%20and%20it%27s%20fantastic%21%20It%27s%20free%20for%20OSS%20and%20offers%20a%20free%20trial%20for%20the%20proprietary%20code.%20Check%20it%20out%3A%20https%3A%2F%2Fcoderabbit.ai)
- [Reddit](https://www.reddit.com/submit?title=Great%20tool%20for%20code%20review%20-%20CodeRabbit&text=I%20just%20used%20CodeRabbit%20for%20my%20code%20review%2C%20and%20it%27s%20fantastic%21%20It%27s%20free%20for%20OSS%20and%20offers%20a%20free%20trial%20for%20proprietary%20code.%20Check%20it%20out%3A%20https%3A//coderabbit.ai)
- [LinkedIn](https://www.linkedin.com/sharing/share-offsite/?url=https%3A%2F%2Fcoderabbit.ai&mini=true&title=Great%20tool%20for%20code%20review%20-%20CodeRabbit&summary=I%20just%20used%20CodeRabbit%20for%20my%20code%20review%2C%20and%20it%27s%20fantastic%21%20It%27s%20free%20for%20OSS%20and%20offers%20a%20free%20trial%20for%20proprietary%20code)

</details>

<sub>Comment `@coderabbitai help` to get the list of available commands and usage tips.</sub>

**Comment by orivaris** [comment]
**coderabbitai[bot]** submitted review **COMMENTED** on 2026-03-17T23:26:12Z:

---

**Actionable comments posted: 1**

<details>
<summary>🧹 Nitpick comments (1)</summary><blockquote>

<details>
<summary>modules/mini-chat/mini-chat/src/domain/service/attachment_service.rs (1)</summary><blockquote>

`435-475`: **Keep provider-consistency checks in the race fallback too.**

`handle_vector_store_insert_race` can return/poll on an existing row without validating `row.provider` against `expected_backend`, unlike the fast path. Reusing the same guard here keeps behavior consistent.


<details>
<summary>♻️ Suggested refactor</summary>

```diff
-            Err(e) => {
-                self.handle_vector_store_insert_race(&conn, scope, chat_id, e)
+            Err(e) => {
+                self.handle_vector_store_insert_race(&conn, scope, chat_id, &expected_backend, e)
                     .await
             }
@@
     async fn handle_vector_store_insert_race(
         &self,
         conn: &modkit_db::DbConn<'_>,
         scope: &AccessScope,
         chat_id: Uuid,
+        expected_backend: &str,
         original_err: DomainError,
     ) -> Result<String, DomainError> {
@@
-            Ok(Some(row)) if row.vector_store_id.is_some() => {
+            Ok(Some(row)) if row.provider != expected_backend => Err(DomainError::conflict(
+                "provider_mismatch",
+                format!(
+                    "vector store provider mismatch: existing='{}', current='{expected_backend}'",
+                    row.provider
+                ),
+            )),
+            Ok(Some(row)) if row.vector_store_id.is_some() => {
                 tracing::warn!(
                     chat_id = %chat_id,
                     error = %original_err,
                     "vector store insert failed but row exists (concurrent insert); using existing"
                 );
                 #[allow(clippy::unwrap_used)]
                 Ok(row.vector_store_id.unwrap())
             }
```
</details>

<details>
<summary>🤖 Prompt for AI Agents</summary>

```
Verify each finding against the current code and only fix it if needed.

In `@modules/mini-chat/mini-chat/src/domain/service/attachment_service.rs` around
lines 435 - 475, handle_vector_store_insert_race currently accepts an existing
row without verifying the provider; update it to enforce the same
provider-consistency as the fast path by comparing row.provider to the
expected_backend before reusing or polling: in the Ok(Some(row)) branch that
returns row.vector_store_id, check row.provider == expected_backend and return
an error (or propagate original_err) on mismatch; likewise in the Ok(Some(_))
placeholder/polling branch, verify the stored row's provider matches
expected_backend before calling poll_vector_store(scope, chat_id). Use the
existing vector_store_repo.find_by_chat result, the expected_backend identifier
from the calling context, and the poll_vector_store and
handle_vector_store_insert_race functions to implement these checks.
```

</details>

</blockquote></details>

</blockquote></details>

<details>
<summary>🤖 Prompt for all review comments with AI agents</summary>

```
Verify each finding against the current code and only fix it if needed.

Inline comments:
In `@modules/mini-chat/mini-chat/src/domain/error.rs`:
- Around line 210-214: The current branch that handles
modkit_db::secure::is_unique_violation(db_err) returns DomainError::Conflict
with db_err.to_string(), which may leak DB internals; instead, log the full
db_err (e.g., using the existing logger or tracing::error!) and return a
sanitized message in DomainError::Conflict such as a generic "unique constraint
violation" or "resource already exists" while keeping db_err only in logs;
update the handling inside the function that constructs DomainError::Conflict
(referencing modkit_db::secure::is_unique_violation, DomainError::Conflict, and
db_err) to perform logging then return the sanitized client-facing message.

---

Nitpick comments:
In `@modules/mini-chat/mini-chat/src/domain/service/attachment_service.rs`:
- Around line 435-475: handle_vector_store_insert_race currently accepts an
existing row without verifying the provider; update it to enforce the same
provider-consistency as the fast path by comparing row.provider to the
expected_backend before reusing or polling: in the Ok(Some(row)) branch that
returns row.vector_store_id, check row.provider == expected_backend and return
an error (or propagate original_err) on mismatch; likewise in the Ok(Some(_))
placeholder/polling branch, verify the stored row's provider matches
expected_backend before calling poll_vector_store(scope, chat_id). Use the
existing vector_store_repo.find_by_chat result, the expected_backend identifier
from the calling context, and the poll_vector_store and
handle_vector_store_insert_race functions to implement these checks.
```

</details>

---

<details>
<summary>ℹ️ Review info</summary>

<details>
<summary>⚙️ Run configuration</summary>

**Configuration used**: defaults

**Review profile**: CHILL

**Plan**: Pro

**Run ID**: `fad0669c-b86e-4cfc-a558-4513ba83fcf2`

</details>

<details>
<summary>📥 Commits</summary>

Reviewing files that changed from the base of the PR and between 77fe61e3ea593898e363967b5120250731a7f607 and b3ce0ce0ebf734d800847cf1c898c87baac3a00e.

</details>

<details>
<summary>⛔ Files ignored due to path filters (1)</summary>

* `dylint_lints/Cargo.lock` is excluded by `!**/*.lock`

</details>

<details>
<summary>📒 Files selected for processing (4)</summary>

* `libs/modkit-db/src/secure/error.rs`
* `libs/modkit-db/src/secure/mod.rs`
* `modules/mini-chat/mini-chat/src/domain/error.rs`
* `modules/mini-chat/mini-chat/src/domain/service/attachment_service.rs`

</details>

</details>

**Comment by orivaris** [comment]
**coderabbitai[bot]** submitted review **COMMENTED** on 2026-03-17T23:33:25Z:

---



<details>
<summary>🧹 Nitpick comments (1)</summary><blockquote>

<details>
<summary>modules/mini-chat/mini-chat/src/domain/service/attachment_service.rs (1)</summary><blockquote>

`452-475`: **Improve fallback observability by separating lookup miss vs lookup error.**

Line 474 currently conflates `Ok(None)` and `Err(_)`, which drops useful diagnostics when the fallback read fails.


<details>
<summary>🔧 Suggested refactor</summary>

```diff
-        match self
-            .vector_store_repo
-            .find_by_chat(conn, scope, chat_id)
-            .await
-        {
+        match self.vector_store_repo.find_by_chat(conn, scope, chat_id).await {
             Ok(Some(row)) if row.vector_store_id.is_some() => {
                 tracing::warn!(
                     chat_id = %chat_id,
                     error = %original_err,
                     "vector store insert failed but row exists (concurrent insert); using existing"
                 );
                 #[allow(clippy::unwrap_used)]
                 Ok(row.vector_store_id.unwrap())
             }
             Ok(Some(_)) => {
                 tracing::warn!(
                     chat_id = %chat_id,
                     error = %original_err,
                     "vector store insert failed but placeholder exists (concurrent insert); polling"
                 );
                 self.poll_vector_store(scope, chat_id).await
             }
-            _ => Err(original_err),
+            Ok(None) => Err(original_err),
+            Err(find_err) => {
+                tracing::warn!(
+                    chat_id = %chat_id,
+                    insert_error = %original_err,
+                    lookup_error = %find_err,
+                    "vector store insert fallback lookup failed; returning original insert error"
+                );
+                Err(original_err)
+            }
         }
```
</details>

<details>
<summary>🤖 Prompt for AI Agents</summary>

```
Verify each finding against the current code and only fix it if needed.

In `@modules/mini-chat/mini-chat/src/domain/service/attachment_service.rs` around
lines 452 - 475, The fallback match currently treats lookup misses and lookup
errors the same; change the match on self.vector_store_repo.find_by_chat(conn,
scope, chat_id).await so Ok(Some(row)) with Some id and Ok(Some(_)) remain
as-is, but split the final arm into Ok(None) => {
tracing::warn!(chat_id=%chat_id, "vector store insert failed and lookup returned
none (no row)"; return Err(original_err) } and Err(lookup_err) => {
tracing::error!(chat_id=%chat_id, error=%lookup_err, "vector store insert failed
and lookup itself errored"); return Err(original_err) } (or similar logging) so
poll_vector_store and error handling in poll_vector_store, original_err,
vector_store_repo.find_by_chat, poll_vector_store, chat_id, scope, conn are all
the referenced symbols to locate and update.
```

</details>

</blockquote></details>

</blockquote></details>

<details>
<summary>🤖 Prompt for all review comments with AI agents</summary>

```
Verify each finding against the current code and only fix it if needed.

Nitpick comments:
In `@modules/mini-chat/mini-chat/src/domain/service/attachment_service.rs`:
- Around line 452-475: The fallback match currently treats lookup misses and
lookup errors the same; change the match on
self.vector_store_repo.find_by_chat(conn, scope, chat_id).await so Ok(Some(row))
with Some id and Ok(Some(_)) remain as-is, but split the final arm into Ok(None)
=> { tracing::warn!(chat_id=%chat_id, "vector store insert failed and lookup
returned none (no row)"; return Err(original_err) } and Err(lookup_err) => {
tracing::error!(chat_id=%chat_id, error=%lookup_err, "vector store insert failed
and lookup itself errored"); return Err(original_err) } (or similar logging) so
poll_vector_store and error handling in poll_vector_store, original_err,
vector_store_repo.find_by_chat, poll_vector_store, chat_id, scope, conn are all
the referenced symbols to locate and update.
```

</details>

---

<details>
<summary>ℹ️ Review info</summary>

<details>
<summary>⚙️ Run configuration</summary>

**Configuration used**: defaults

**Review profile**: CHILL

**Plan**: Pro

**Run ID**: `9ed29a35-2207-4a90-b5c2-eae2b232917a`

</details>

<details>
<summary>📥 Commits</summary>

Reviewing files that changed from the base of the PR and between b3ce0ce0ebf734d800847cf1c898c87baac3a00e and 3b45986b05deaa5bf5bbfaa72ced371ed6296df0.

</details>

<details>
<summary>⛔ Files ignored due to path filters (1)</summary>

* `dylint_lints/Cargo.lock` is excluded by `!**/*.lock`

</details>

<details>
<summary>📒 Files selected for processing (4)</summary>

* `libs/modkit-db/src/secure/error.rs`
* `libs/modkit-db/src/secure/mod.rs`
* `modules/mini-chat/mini-chat/src/domain/error.rs`
* `modules/mini-chat/mini-chat/src/domain/service/attachment_service.rs`

</details>

<details>
<summary>🚧 Files skipped from review as they are similar to previous changes (2)</summary>

* modules/mini-chat/mini-chat/src/domain/error.rs
* libs/modkit-db/src/secure/mod.rs

</details>

</details>

**Comment by orivaris** [comment]
**coderabbitai[bot]** submitted review **COMMENTED** on 2026-03-18T00:17:02Z:

---



<details>
<summary>♻️ Duplicate comments (1)</summary><blockquote>

<details>
<summary>modules/mini-chat/mini-chat/src/domain/error.rs (1)</summary><blockquote>

`210-214`: _⚠️ Potential issue_ | _🟠 Major_

**Sanitize the fallback 409 body.**

Line 213 still sends `db_err.to_string()` back to clients, so wrapped driver errors can leak constraint or table details through the conflict response.
 

<details>
<summary>🔧 Suggested fix</summary>

```diff
     if modkit_db::secure::is_unique_violation(db_err) {
+        tracing::debug!(error = %db_err, "classified unique violation via fallback matcher");
         return DomainError::Conflict {
             code: "unique_violation".into(),
-            message: db_err.to_string(),
+            message: "A conflicting resource already exists".into(),
         };
     }
```
</details>

<details>
<summary>🤖 Prompt for AI Agents</summary>

```
Verify each finding against the current code and only fix it if needed.

In `@modules/mini-chat/mini-chat/src/domain/error.rs` around lines 210 - 214, The
Conflict branch in the is_unique_violation handling is leaking raw DB error text
via db_err.to_string(); update the DomainError::Conflict construction in the
modkit_db::secure::is_unique_violation branch to return a sanitized, generic
message (e.g., "resource conflict" or include only the constraint name if
safe-validated) instead of db_err.to_string(), so sensitive
constraint/table/driver details are not exposed to clients; ensure changes are
made where DomainError::Conflict is created in error.rs.
```

</details>

</blockquote></details>

</blockquote></details>

<details>
<summary>🤖 Prompt for all review comments with AI agents</summary>

```
Verify each finding against the current code and only fix it if needed.

Duplicate comments:
In `@modules/mini-chat/mini-chat/src/domain/error.rs`:
- Around line 210-214: The Conflict branch in the is_unique_violation handling
is leaking raw DB error text via db_err.to_string(); update the
DomainError::Conflict construction in the modkit_db::secure::is_unique_violation
branch to return a sanitized, generic message (e.g., "resource conflict" or
include only the constraint name if safe-validated) instead of
db_err.to_string(), so sensitive constraint/table/driver details are not exposed
to clients; ensure changes are made where DomainError::Conflict is created in
error.rs.
```

</details>

---

<details>
<summary>ℹ️ Review info</summary>

<details>
<summary>⚙️ Run configuration</summary>

**Configuration used**: defaults

**Review profile**: CHILL

**Plan**: Pro

**Run ID**: `b7398821-28f8-4536-bb63-fb18ff2b095d`

</details>

<details>
<summary>📥 Commits</summary>

Reviewing files that changed from the base of the PR and between 3b45986b05deaa5bf5bbfaa72ced371ed6296df0 and aabd0e040f7db9eab06039a8c8cb047d6c721bd6.

</details>

<details>
<summary>⛔ Files ignored due to path filters (1)</summary>

* `dylint_lints/Cargo.lock` is excluded by `!**/*.lock`

</details>

<details>
<summary>📒 Files selected for processing (4)</summary>

* `libs/modkit-db/src/secure/error.rs`
* `libs/modkit-db/src/secure/mod.rs`
* `modules/mini-chat/mini-chat/src/domain/error.rs`
* `modules/mini-chat/mini-chat/src/domain/service/attachment_service.rs`

</details>

<details>
<summary>🚧 Files skipped from review as they are similar to previous changes (1)</summary>

* libs/modkit-db/src/secure/error.rs

</details>

</details>

**Comment by orivaris** [comment]
**coderabbitai[bot]** reviewed `modules/mini-chat/mini-chat/src/domain/error.rs` line 214 on 2026-03-17T23:26:11Z:

---

_⚠️ Potential issue_ | _🟠 Major_

**Avoid exposing raw DB error strings in 409 responses.**

Line 213 currently returns `db_err.to_string()` to clients via `DomainError::Conflict`. This can leak internal database details. Keep full error in logs, but return a sanitized conflict message.


<details>
<summary>🔧 Suggested fix</summary>

```diff
     if modkit_db::secure::is_unique_violation(db_err) {
+        tracing::debug!(error = %db_err, "classified unique violation via fallback matcher");
         return DomainError::Conflict {
             code: "unique_violation".into(),
-            message: db_err.to_string(),
+            message: "A conflicting resource already exists".into(),
         };
     }
```
</details>

<details>
<summary>🤖 Prompt for AI Agents</summary>

```
Verify each finding against the current code and only fix it if needed.

In `@modules/mini-chat/mini-chat/src/domain/error.rs` around lines 210 - 214, The
current branch that handles modkit_db::secure::is_unique_violation(db_err)
returns DomainError::Conflict with db_err.to_string(), which may leak DB
internals; instead, log the full db_err (e.g., using the existing logger or
tracing::error!) and return a sanitized message in DomainError::Conflict such as
a generic "unique constraint violation" or "resource already exists" while
keeping db_err only in logs; update the handling inside the function that
constructs DomainError::Conflict (referencing
modkit_db::secure::is_unique_violation, DomainError::Conflict, and db_err) to
perform logging then return the sanitized client-facing message.
```

</details>

**Comment by orivaris** [comment]
**MikeFalcon77** reviewed `modules/mini-chat/mini-chat/src/domain/error.rs` line 214 on 2026-03-18T00:07:13Z:

---

fixed

**Comment by orivaris** [comment]
**coderabbitai[bot]** reviewed `modules/mini-chat/mini-chat/src/domain/error.rs` line 214 on 2026-03-18T00:07:54Z:

---

<details>
<summary>🧩 Analysis chain</summary>

🏁 Script executed:
```
#!/bin/bash
# Check the current state of map_db_err in the PR branch
grep -n -A 20 "fn map_db_err" modules/mini-chat/mini-chat/src/domain/error.rs
```

Length of output: 789

---



</details>

`@MikeFalcon77`, the branch still shows the original code at Line 213 — `message: db_err.to_string()` is still being returned directly in the fallback `is_unique_violation` path, and no debug log was added. It looks like the fix may not have been pushed yet. Here's the change needed:

```rust
if modkit_db::secure::is_unique_violation(db_err) {
    tracing::debug!(error = %db_err, "classified unique violation via fallback matcher");
    return DomainError::Conflict {
        code: "unique_violation".into(),
        message: "A conflicting resource already exists".into(),
    };
}
```

Note that the first branch (the `sql_err()` path on Line 201–206) passes `msg` from SeaORM's structured `UniqueConstraintViolation` variant — that `msg` is the constraint name, not a raw driver string, so it's generally safe. The concern specifically applies to the fallback path on Line 213 where `db_err.to_string()` can include raw driver-level error text (e.g., Postgres error codes, table/column names, internal constraint detail).

---

<details>
<summary>🧠 Learnings used</summary>

```
Learnt from: aviator5
Repo: cyberfabric/cyberfabric-core PR: 1011
File: modules/mini-chat/mini-chat/src/infra/oagw_provisioning.rs:37-44
Timestamp: 2026-03-11T12:11:58.818Z
Learning: In `modules/mini-chat/mini-chat/src/infra/oagw_provisioning.rs`, OAGW does not persist upstreams to a database, so duplicate upstream creation cannot occur on restart. Therefore, `create_upstream` returning `None` always signals a genuine failure, and the fail-fast `ok_or_else(...)?` pattern in `register_oagw_upstreams` is correct. Do not revert to warn-and-continue on this basis — the previous idempotent-startup concern no longer applies.
```

---

```
Learnt from: striped-zebra-dev
Repo: cyberfabric/cyberfabric-core PR: 908
File: libs/modkit-db/src/outbox/sequencer.rs:118-127
Timestamp: 2026-03-09T11:24:39.573Z
Learning: In `libs/modkit-db/src/outbox/sequencer.rs`, the `claim_incoming_for_partition()` function's SQLite path performs a SELECT-then-DELETE without verifying the DELETE row count. This is intentional by design: SQLite is documented as single-process only (`Dialect::Sqlite` doc: "Do not run multiple outbox instances against the same SQLite database"), and SQLite's exclusive write locking within a transaction guarantees that no concurrent writer can claim or delete the same rows. Do not flag the missing `ExecResult`/`rows_affected` check in the SQLite claim path as a race condition or correctness issue.
```

---

```
Learnt from: striped-zebra-dev
Repo: cyberfabric/cyberfabric-core PR: 1094
File: modules/mini-chat/mini-chat/src/domain/service/attachment_service_test.rs:499-537
Timestamp: 2026-03-13T20:54:01.297Z
Learning: In `modules/mini-chat/mini-chat/src/domain/service/attachment_service_test.rs`, the test `test_vector_store_failure_sets_attachment_failed` (lines 878-934) intentionally uses `count_ready_documents() == 0` as its primary DB assertion, combined with an error-message check. The author acknowledges this is a weak DB state check (it cannot distinguish Failed from Pending/Uploaded) but considers the combined error-return + error-message coverage sufficient for P1. Do not re-flag `ready_count == 0` as the sole DB verification if the test also validates the returned error variant and message. A direct `status == Failed` assertion would be strictly stronger but is not required for this PR.
```

---

```
Learnt from: striped-zebra-dev
Repo: cyberfabric/cyberfabric-core PR: 1018
File: modules/mini-chat/mini-chat/src/infra/db/repo/quota_usage_repo.rs:118-123
Timestamp: 2026-03-11T18:46:29.610Z
Learning: In `modules/mini-chat/mini-chat/src/infra/db/repo/quota_usage_repo.rs`, the `web_search_delta.cast_signed()` call in `settle()` is intentional and safe by design. `web_search_delta` is sourced from `web_search_completed_count`, which is capped upstream at `max_calls_per_message` (default 2), making values above `i32::MAX` structurally unreachable. This mirrors the same `cast_signed()` pattern used for token deltas in the same file. Do not flag this cast as unsafe or suggest replacing it with `i32::try_from()`.
```

---

```
Learnt from: striped-zebra-dev
Repo: cyberfabric/cyberfabric-core PR: 1018
File: modules/mini-chat/mini-chat/src/infra/db/repo/quota_usage_repo.rs:195-217
Timestamp: 2026-03-11T16:58:24.317Z
Learning: In `modules/mini-chat/mini-chat/src/infra/db/repo/quota_usage_repo.rs`, `get_daily_web_search_calls` intentionally performs an unlocked (optimistic) read. Per DESIGN.md §5.8 (line 2579), preflight quota checks are admission control only — not transactional guarantees. Pessimistic locking (`FOR UPDATE`) is used only during settlement (`find_bucket_rows_for_update`). The hard enforcement backstop is the per-message `max_calls_per_message` limit enforced mid-stream. Do not flag the unlocked preflight read as a race condition or suggest adding `SELECT … FOR UPDATE` at preflight.
```

---

```
Learnt from: striped-zebra-dev
Repo: cyberfabric/cyberfabric-core PR: 906
File: modules/mini-chat/mini-chat/src/domain/repos/outbox_enqueuer.rs:50-54
Timestamp: 2026-03-06T13:18:58.731Z
Learning: In `modules/mini-chat/mini-chat/src/domain/repos/outbox_enqueuer.rs`, the `OutboxEnqueuer` trait's generic method `enqueue_usage_event<C: DBRunner>` (which makes it non-object-safe) is intentional scaffolding. The wiring of `OutboxEnqueuer` into `FinalizationService` (as `Arc<dyn OutboxEnqueuer>` or via generics) is explicitly deferred until `modkit_db::outbox` lands. TODOs in the code (labelled P4) and tasks.md track this. Do not flag the non-object-safe generic or the missing wiring as a bug in the interim.
```

---

```
Learnt from: striped-zebra-dev
Repo: cyberfabric/cyberfabric-core PR: 998
File: modules/mini-chat/mini-chat/src/domain/repos/message_repo.rs:121-160
Timestamp: 2026-03-10T19:18:46.084Z
Learning: In `modules/mini-chat/mini-chat/src/domain/repos/message_repo.rs`, the dual use of `None` in the snapshot-boundary API is intentional and correct by design: `snapshot_boundary()` returns `None` only when the chat has zero messages, and `recent_for_context`/`recent_after_boundary` treat `None` as "no upper bound" — but the chat-scoped filters mean zero-message chats always return zero rows anyway. Retry/edit flows never produce a `None` boundary because messages are not soft-deleted. Do not flag `Option<SnapshotBoundary>` as conflating "empty history" and "unbounded" — the two cases are semantically equivalent in practice.
```

---

```
Learnt from: mattgarmon
Repo: cyberfabric/cyberfabric-core PR: 1160
File: modules/system/oagw/oagw/src/domain/services/management.rs:570-625
Timestamp: 2026-03-16T21:50:45.202Z
Learning: In `modules/system/oagw/oagw/src/domain/services/management.rs`, the `validate_route_determinism` function performs a non-atomic read-before-write check (via `routes.list`) for route match conflicts. This is intentionally a temporary service-layer guard only; the authoritative uniqueness enforcement will be added at the repository/storage layer (e.g., DB unique constraint or transactional upsert) during the upcoming database integration work. Do not re-flag this service-layer check as a critical race condition bug — the team is aware and deferring the atomic fix to the database integration milestone.
```

---

```
Learnt from: aviator5
Repo: cyberfabric/cyberfabric-core PR: 1034
File: modules/mini-chat/mini-chat/src/module.rs:13-13
Timestamp: 2026-03-12T12:20:14.216Z
Learning: In `modules/mini-chat/mini-chat/src/module.rs`, the `exchange_client_credentials` function (around line 361) needlessly calls `creds.client_secret.expose_secret().to_owned()` and re-wraps the result in a `SecretString` when constructing `ClientCredentialsRequest`, instead of simply using `creds.client_secret.clone()`. The `use secrecy::ExposeSecret` import at line 13 is also unnecessary as a result. The fix (use `creds.client_secret.clone()` and remove the `ExposeSecret` import) is explicitly deferred to a follow-up PR after PR `#1034`. Do not re-flag this secret declassification until the follow-up PR lands.
```

---

```
Learnt from: striped-zebra-dev
Repo: cyberfabric/cyberfabric-core PR: 908
File: libs/modkit-db/src/outbox/strategy.rs:346-372
Timestamp: 2026-03-09T11:25:40.297Z
Learning: In `libs/modkit-db/src/outbox/strategy.rs` (DecoupledStrategy), `lease_release()` in `dialect.rs` only clears `locked_by` and `locked_until` — it does NOT decrement or reset the `attempts` counter that `lease_acquire()` increments. Empty polls that commit the lease-acquire transaction and then call `lease_release()` permanently inflate `attempts`. The correct fix for the empty-path is to `txn.rollback()` before returning `Ok(None)`, so the `attempts` increment is never persisted.
```

---

```
Learnt from: MikeFalcon77
Repo: cyberfabric/cyberfabric-core PR: 1181
File: modules/mini-chat/mini-chat/src/domain/citation_mapping.rs:27-28
Timestamp: 2026-03-16T23:15:21.719Z
Learning: In `modules/mini-chat/mini-chat/src/domain/citation_mapping.rs`, `c.title` on a `Citation` with `CitationSource::File` originates from the Azure OpenAI file_search provider citation response, not from direct user input. It mirrors the document filename already stored in the DB. Per the project's security guidelines, filenames are not sensitive (only secrets, tokens, and credentials are). Logging `title = %c.title` in the malformed-citation `tracing::warn!` is intentional for debuggability and should not be flagged as a sensitive-data leak.
```

---

```
Learnt from: striped-zebra-dev
Repo: cyberfabric/cyberfabric-core PR: 908
File: libs/modkit-db/src/outbox/processor.rs:0-0
Timestamp: 2026-03-07T20:03:14.605Z
Learning: In `libs/modkit-db/src/outbox/processor.rs`, the `attempts` field stored in a dead-letter row (`insert_dead_letter`) intentionally uses the pre-increment value (`msg.attempts`), because `attempts` is incremented *after* the handler returns. The dead-letter record captures how many attempts were made *before* the current (terminal) rejection. This is validated by the `crash_leaves_incremented_attempts_in_db` test. Do not flag this as an off-by-one error.
```

---

```
Learnt from: striped-zebra-dev
Repo: cyberfabric/cyberfabric-core PR: 1094
File: modules/mini-chat/mini-chat/src/domain/service/context_assembly.rs:68-94
Timestamp: 2026-03-13T20:51:55.573Z
Learning: In `modules/mini-chat/mini-chat/src/domain/service/context_assembly.rs`, the `ContextAssemblyError::BudgetExceeded` Display message ("mandatory context items require N tokens but only M are available") is intentionally generic. The error is internal-only; HTTP clients receive the structured `context_length_exceeded` error code (DESIGN.md line 843). When `max_output_tokens_applied` or tool surcharges alone exhaust the context window, `required_tokens` carries the full deduction total (output + surcharges + overhead), so the message is technically accurate — reservations are treated as mandatory items. Do not flag this wording as misleading or suggest adding a reason/variant to distinguish reservation exhaustion from prompt-item exhaustion.
```

---

```
Learnt from: aviator5
Repo: cyberfabric/cyberfabric-core PR: 926
File: modules/mini-chat/mini-chat/src/infra/db/migrations/m20260302_000001_initial.rs:38-38
Timestamp: 2026-03-07T10:37:11.709Z
Learning: The mini-chat module in `cyberfabric/cyberfabric-core` has no existing production or customer installations as of PR `#926` (March 2026). It is under active development. Schema changes (new tables, indexes) may be added directly to the initial migration file (`modules/mini-chat/mini-chat/src/infra/db/migrations/m20260302_000001_initial.rs`) rather than in a separate incremental migration, because there are no deployed databases to upgrade. Do not flag amendments to the initial migration as a critical upgrade-path bug until the module has declared a stable/released state.
```

---

```
Learnt from: striped-zebra-dev
Repo: cyberfabric/cyberfabric-core PR: 1094
File: modules/mini-chat/mini-chat/src/domain/service/attachment_service_test.rs:938-991
Timestamp: 2026-03-13T20:51:25.357Z
Learning: In `modules/mini-chat/mini-chat/src/domain/service/attachment_service_test.rs`, the test `test_upload_concurrent_delete_during_upload` intentionally does NOT call `upload_file()` or force `cas_set_uploaded()` to return 0. It only verifies that a pre-deleted attachment row is unreadable via `get_attachment()` (soft-delete filtering boundary). True CAS=0 concurrent testing is impractical under SQLite's single-writer model. The CAS guard pattern is covered by turn finalization CAS tests. Do not flag this test as failing to exercise the concurrent-delete path; the inline comments (lines 949-964) document the rationale.
```

---

```
Learnt from: striped-zebra-dev
Repo: cyberfabric/cyberfabric-core PR: 940
File: modules/mini-chat/mini-chat/src/api/rest/handlers/messages.rs:73-77
Timestamp: 2026-03-07T18:15:02.690Z
Learning: In `cyberfabric/cyberfabric-core` mini-chat module, `chat.model` is immutable after chat creation and does not change between turns. In `modules/mini-chat/mini-chat/src/api/rest/handlers/messages.rs`, `selected_model` is correctly sourced from `chat.model` as cosmetic metadata for SSE Done events and replay. The replay path (`replay::replay_turn`) reads `effective_model` directly from the persisted `TurnModel` record for downgrade metadata, so there is no stale-model risk if `selected_model` is taken from the current chat. Do not flag this pattern as a replay correctness bug.
```

---

```
Learnt from: aviator5
Repo: cyberfabric/cyberfabric-core PR: 847
File: modules/mini-chat/mini-chat/src/domain/service/mod.rs:40-42
Timestamp: 2026-03-02T17:22:47.283Z
Learning: Enforce the GTS naming convention across Rust source files: use ~ as a delimiter where type definitions end with a trailing ~ (e.g., gts.x.core.users.user.v1~), instances separate type from instance with ~ (e.g., gts.acme.core.models.user.v1~acme.core.instances.user1.v1), and compound/aliased types chain multiple identifiers with ~ (e.g., gts.x.core.lic.feat.v1~x.core.global.base.v1). Do not flag a trailing or intermediate ~ as an error; treat it as intentional syntax. Apply these checks to Rust files (*.rs) throughout the codebase.
```

---

```
Learnt from: striped-zebra-dev
Repo: cyberfabric/cyberfabric-core PR: 1094
File: modules/mini-chat/mini-chat/src/domain/service/attachment_service_test.rs:860-872
Timestamp: 2026-03-13T20:51:12.091Z
Learning: In the mini-chat test suite (e.g., modules/mini-chat/mini-chat/src/domain/service/attachment_service_test.rs and related tests), the pattern tokio::time::sleep(std::time::Duration::from_millis(50)) is used to gate assertions on fire-and-forget tokio::spawn tasks. These spawned tasks perform only instant/mock provider calls and in-memory writes, so a 50ms sleep is far larger than real execution. Do not flag these fixed sleeps as timing-dependent flakes in this context, and avoid suggesting removal unless a formal flake-detection mechanism in CI explicitly covers this scenario.
```

---

```
Learnt from: striped-zebra-dev
Repo: cyberfabric/cyberfabric-core PR: 1094
File: modules/mini-chat/mini-chat/src/domain/repos/vector_store_repo.rs:40-45
Timestamp: 2026-03-13T20:53:30.975Z
Learning: In the cyberfabric/cyberfabric-core Mini-chat module, all repository trait methods must rely on tenant isolation via the ModKit SecureORM pattern. Use scope: &AccessScope with .secure().scope_with(scope); the AccessScope carries tenant_id and is applied automatically as a WHERE clause. Do not require or flag explicit tenant_id parameters on repo methods; the SecureORM machinery enforces isolation. Treat missing explicit tenant_id as not a gap, assuming CI/tests validate correct scope usage. This guidance applies to all repo methods within this module (and similar code paths using SecureORM) and should be checked by code reviews and automated validation.
```

---

```
Learnt from: striped-zebra-dev
Repo: cyberfabric/cyberfabric-core PR: 1155
File: modules/mini-chat/mini-chat/src/domain/stream_events.rs:22-27
Timestamp: 2026-03-16T14:14:32.369Z
Learning: In the mini-chat domain, ensure the DESIGN.md SSE contract referenced by changes to stream_events.rs defines: non-terminal events exactly as [turn_started, delta, tool, citations, ping], terminal events as [done, error], and the normative ordering grammar: turn_started ping* (delta | tool)* citations? (done | error). The event model should total seven event types. Do not flag turn_started as missing from the DESIGN.md contract. Use this as a review checklist when touching modules/mini-chat/mini-chat/src/domain/stream_events.rs and related files.
```

---

```
Learnt from: Artifizer
Repo: cyberfabric/cyberfabric-core PR: 949
File: dylint_lints/de09_gts_layer/de0901_gts_string_pattern/src/lib.rs:58-65
Timestamp: 2026-03-17T21:48:17.987Z
Learning: Global review guidance: The check_item function enforcing the _WILDCARD suffix convention for const/static GTS wildcard strings in dylint_lints/de09_gts_layer/de0901_gts_string_pattern/src/lib.rs should be treated as a cross-crate, cross-module convention and not limited to contract modules (i.e., do not scope it only to is_in_contract_module_ast). When reviewing similar code elsewhere, apply this naming convention check universally across all modules and crates unless there is an explicit, automated scope gate to restrict it.
```

</details>
